[RNMobile] Add "Set as Featured Image" Button to Image Block (Android Only)

SiobhyB opened the pull request:
Partial fix for: https://github.com/wordpress-mobile/gutenberg-mobile/issues/1011

`gutenberg-mobile`: https://github.com/wordpress-mobile/gutenberg-mobile/pull/3116
`WordPress-Android`: https://github.com/wordpress-mobile/WordPress-Android/pull/14503

<!-- Learn the overall process and best practices for pull requests at https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/repository-management.md#pull-requests. -->

## Description

Building on the work done in https://github.com/WordPress/gutenberg/pull/30806, this PR will add a **Set as Featured** button to the image block's settings, with the purpose being to make it simpler for users to set a featured image within the post's editor. Users will also be able to **Remove as Featured** directly from the block's settings. 

_Note: This PR introduces the **Set as Featured** button to Android devices only, support for iOS will follow in future iterations, too._

## How has this been tested?

### Test Case 1: General Flow

<details>
<summary>To start with, the general flow for setting a featured image via the image block on a post without an existing image is as follows.</summary>

* In the app, navigate to _Home_ → _Blog Posts_ to create a new blog post, one which doesn't already have a featured image set.
* Add a new image block, and upload an image to it.
* Tap the image block's cog/gear icon to access its settings.
* Observe the new **Set as Featured** button at the bottom of the settings and tap it.
* Verify that you see a confirmation message about the newly set featured image and that the block now has a **Featured** banner overlaying it.
</details>

### Test Case 2: "Replace Existing" Flow

<details>
<summary>It's also possible for users to replace an existing featured image via the following flow.</summary>

* Add a new image block to the post and upload a second image to it.
* Again, tap the image block's cog/gear icon to access its settings.
* Tap the **Set as Featured** button at the bottom of the settings and tap it.
* A dialog should appear to inform you that the post already has a featured image, with an option to **Cancel** or confirm your wish to **Replace**. 
* Tap **Cancel** to confirm it works as expected, to dismiss the action.
* Go through steps two through four again and this time tap **Replace**.
* Verify that you see a confirmation message about the newly set featured image and that the block now has a **Featured** banner overlaying it. The banner should no longer overlay the previous featured image. </details>

### Test Case 3: "Editor Loads" Flow

<details><summary>Next, we should verify that a **Remove as Featured** option appears when a post with a featured image is first loaded. To test this, follow these steps.</summary>

* Verify that a featured image has been set on the post and tap the back button.
* Return to the post's editor by tapping on it from the post list screen.
* Verify that the post's featured image has a **Featured** banner overlaying it. Tap the cog/gear icon to then confirm the option to **Remove as Featured** appears.</details>

### Test Case 4: "Post Settings" Flow

<details><summary>The image block should be updated to reflect whether its image is featured or not even when the featured image is set via the general Post Settings flow.</summary>

* Tap the three icons to the upper right of the post editor to access **Post Settings**.
* Tap existing featured image and then **Remove**.
* Tap the back button to confirm that the **Featured** banner has been removed from the featured image, as expected. Tap the cog/gear icon to then confirm the option to **Set as Featured** appears.
* Return to the **Post Settings** screen and set a new featured image.
* Tap the back button. If the new featured image hasn't already been added to the editor via an image block, add it.  
* Confirm that the **Featured** banner displays overlays the correct image and that  **Remove as Featured** appears in its settings.</details>

### Test Case 5: "Replace in Image Block" Flow

<details><summary>Finally, it's possible to replace an image directly within the image block's settings. We need to verify that the featured settings update as expected depending on the image that's chosen for an image block.</summary>

* Select any image block that is flagged with the **Featured** banner.
* Tap the edit icon to the upper right and then **Replace**. 
* Replace with any image that _isn't_ the post's featured image.
* Verify that the **Featured** banner is removed from the block when the new image uploads and that the **Set as Featured** button appears in the block's settings.
* Go through the steps to replace the block's image again, this time pick the post's featured image.
* Verify that the **Featured** banner appears again and that the **Remove as Featured** button appears in the block's settings.</details>

## Screenshots <!-- if applicable -->

| Set as Featured Image  | Remove as Featured Image |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/113282237-8fecfc00-92de-11eb-9618-3f84c10e1720.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/113282171-7cda2c00-92de-11eb-90ab-738e9e90591d.png" width="100%"> |

| Replace Dialog  | Confirmation Dialog |
| ------------- | ------------- |
| <img src="https://user-images.githubusercontent.com/2998162/112844989-3f865c00-909c-11eb-9be2-6bfac25b8b60.png" width="100%"> | <img src="https://user-images.githubusercontent.com/2998162/113282281-9da28180-92de-11eb-89f4-9e2b791f84ec.png" width="100%"> |

## Types of changes

This PR introduces a new feature (non-breaking change which adds functionality). The feature it introduces is a **Set as Featured** button in the image block's setting, which enables a user to either set an image as featured or remove it as featured. 

A high-level overview of the code changes involved is as follows:

* [A new component is defined](https://github.com/WordPress/gutenberg/pull/28854/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R442) with `getSetFeaturedButton`. This component outputs a button with either a `Set as Featured` or a `Remove as Featured` label, depending on whether `isFeaturedImage` is true or not.
* `onSetFeatured` will be called when the button is tapped, [this function](https://github.com/WordPress/gutenberg/pull/28854/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R436) utilizes another function named `setFeaturedImage` to [send an image's media ID over the bridge](https://github.com/WordPress/gutenberg/pull/28854/files#diff-aeea93b0485478789972ba668d7057b35b0b366be9ce02de0004b6d0356489bc) when a request is made to either set a new featured image (in which case the media ID is that of the selected image) or to remove an existing featured image (in which case the media ID is zero). A [call to close the settings screen](https://github.com/WordPress/gutenberg/pull/28854/files#diff-7126ae03000a022bcefa0d2a5fb03b6aca56ed4047e3a49d2131babf229cf6c1R439) is also made within `onSetFeatured`.
* Building on the logic for grabbing the ID of a post's featured image, which was set out in https://github.com/WordPress/gutenberg/pull/30806, `featuredImageId` is updated after confirmation of the change on the native side, as native is our main source of truth when it comes to confirming the change to the post's featured image ID. The changes to `packages/edit-post/src/editor.native.js` in the linked PR can be referenced for further details around how this is handled using the `getEditedPostAttribute()` function.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://developer.wordpress.org/coding-standards/wordpress-coding-standards/accessibility/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://developer.wordpress.org/coding-standards/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR. <!-- React Native mobile Gutenberg guidelines: https://github.com/WordPress/gutenberg/blob/HEAD/docs/contributors/native-mobile.md -->

SiobhyB committed:
Add setFeaturedImage() methods to bridge

SiobhyB committed:
Add getSetFeaturedButton() function which returns a "set/remove featured| button

The function returns a button (in the form of a BottomSheet.Cell component) based on whether "isFeaturedImage" is true or false. The function called within the image block's settings.

In its current form, the button is unresponsive and only includes a label to "Set as Featured Image" or "Remove as Featured Image", depending on whether the current image is already featured or not.

SiobhyB committed:
Wrap "androidOnly" conditional around "Set/Remove Featured Image" button

For now, the button will only work on Android devices, so an "androidOnly" conditional is necessary to avoid bugs/complications on iOS.

SiobhyB committed:
Define button's CSS selectors for styling

The getStylesFromColorScheme hook is used to define CSS selectors for the button, depending on whether the device uses dark or light mode. The hook's "featuredButtonStyle" variable is then called via BottomSheet's labelStyle attribute.

SiobhyB committed:
Update styles.native.scss with styling for featuredButton and featuredButtonDark

SiobhyB committed:
Define CSS selector for "Set as Featured Image" button

This commit uses the "getStylesFromColorScheme" hook to set up CSS selectors for the "Set as Featured Image" button, as its style will vary slightly from the base button.

SiobhyB committed:
Update styles.native.scss with styling for setFeaturedButton

This includes CSS for both the light mode and dark mode versions of the button.

SiobhyB committed:
Define CSS selector for "Remove as Featured Image" button

SiobhyB committed:
Update styles.native.scss with styling for removeFeaturedButton

SiobhyB committed:
Set up "onSetFeatured" function for when button is pressed

This function will fire when the "Set/Remove as Featured Image" button is pressed and its functionality will be fleshed out in future commits.

SiobhyB committed:
Import setFeaturedImage and call it from onSetFeatured

setFeaturedImage is imported from '@wordpress/react-native-bridge' and called via the onSetFeatured, as it'll be utilised when the button is pressed.

SiobhyB committed:
Pass mediaId of new featured image into onSetFeatured() and setFeaturedImage()

When a featured image is being removed, the mediaId will be zero, and when it's being set it'll be the "attributes.id" of the image in the block.

SiobhyB committed:
Add closeSettingsBottomSheet() to dismiss BottomSheet when button's pressed

SiobhyB committed:
Update CHANGELOG.md

SiobhyB committed:
Update reference to 'core/edit-post' to avoid store name literal

hypest commented:
> SiobhyB requested a review from hypest 31 minutes ago

👋 @SiobhyB , happy to review! Can you share some info about what's not "finished" yet keeping the PR in Draft mode? Anything in particular you'd like me to offer feedback on? Thanks!

SiobhyB commented:
@hypest, thanks for taking a look! The `gutenberg-mobile` PR still has a couple of failing tests that I'm looking into, but the overall code is functioning. 

I'm keeping the PR as a draft as there will likely be some commits coming as part of the efforts to resolve the failing tests, but I believe that these will mainly be merges and updates to references. I don't currently have any other major changes to the code in mind, so seeking feedback on the overall PRs, aside from the failing tests on `gutenberg-mobile` (unless you notice something in the code that's causing them to fail, of course!). Looking forward to hearing what you think and thanks again!

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB commented:
@hypest, just an FYI that I'm taking this out of draft now as all tests have now passed. :D 🎉

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -487,6 +547,7 @@ export class ImageEdit extends Component {
 				<PanelBody title={ __( 'Link Settings' ) }>
 					{ this.getLinkSettings( true ) }
 				</PanelBody>
+				{ this.getSetFeaturedButton( isFeaturedImage ) }
```
👋 @SiobhyB , what do you think about putting the `androidOnly` conditional here ([example](https://github.com/WordPress/gutenberg/blob/5dec957c24f9afcb5b3e318e179a4cfffa309692/packages/block-library/src/social-links/edit.js#L219)) to avoid the whole running of `getSetFeaturedButton()` on iOS? That way, getSetFeaturedButton get's sligthly cleaner and we also don't run all of its code (like getting styles) unless it will indeed be used (on Android).

SiobhyB replied:
Ah, good point! That makes complete sense and I've gone ahead to commit that change.

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -430,6 +434,62 @@ export class ImageEdit extends Component {
+					androidOnly && (
+						<PanelBody>
+							{ isFeaturedImage ? (
+								<BottomSheet.Cell
```
Nit, and mostly a personal pref but, I wonder what do you think about defining a couple of consts to hold the BottomSheet.Cells and use them in the ternary? I feel like it's a bit too nested and idented the way it currently is, making reading the code more difficult. Again, this comes down to personal preference mostly so, feel free to keep as is.

SiobhyB replied:
I like that idea and agree it'd make the code a bit more readable. :) I've gone ahead to commit it.

hypest commented on packages/block-library/src/image/styles.native.scss:
```diff
@@ -30,3 +30,26 @@
+	border-top-color: $gray-70;
+}
+
+.setFeaturedButton {
```
🤔 I'm a little bit confused by the `featuredButton` vs `setFeaturedButton` distinction. Are there plans for using `featuredButton` in other places and the`setFeaturedButton` style won't be suitable?

If not, what do you think about merging the `featuredButton` and `setFeaturedButton` definitions? In the end, there will be only `setFeaturedButton` and `setFeaturedButtonDark`. I think it won't be a problem that in JS the `color` property will be overwritten for Dark mode and the removeButton case.

SiobhyB replied:
Thanks for the feedback! My hope was that splitting the selectors out in this way, with two separate selectors for the button's two different styles, would make the code more readable. As it seems to have had the oppositive effect, I've rolled `featuredButton` into `setFeaturedButton`, as per your suggestion. :)

hypest commented:
@SiobhyB , I think the [fix for undefined](https://github.com/WordPress/gutenberg/blob/trunk/packages/block-library/src/image/edit.native.js#L451-L453) is not included. Can you do another update from trunk? Thanks!

hypest commented:
I'm getting a red screen with `Attempt to invoke virtual method 'double java.lang.Double.doubleValue()' on a null object reference` when trying to tap on the "Set as Featured Image" on an image in the demo app. Granted, it didn't expect it to work (as in, to indeed set the featured image) but I feel it shouldn't crash either, right? Can you have a look @SiobhyB ? Thanks!

hypest reviewed (commented):
Made a first pass and this works very well @SiobhyB ! 👏

I've left a few comments, most not too important and one about updating from trunk and looking into a crash in the demo app.

I will continue with having a look at the WPAndroid PR as well.

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB committed:
Move androidOnly conditional

The conditional is moved to the main call for this.getSetFeaturedButton(), to prevent unecessary code from being load.

SiobhyB committed:
Create "setFeaturedButton" and "removeFeaturedButton" consts

These consts are then returned by the component, with the purpose being to make the return statement more readable.

SiobhyB committed:
Merge "featuredButton" and "setFeaturedButton" CSS styles

For better readability, this commit merges the separates selectors into one main "setFeaturedButton" selector.

SiobhyB committed:
Add check for whether image's attribute.id is "undefined"

This check will prevent getSetFeaturedButton() running in cases where an image's attribute.id is undefined, as attempting to set an image without a valid id as feature would lead to a crash.

For more background, see: https://github.com/WordPress/gutenberg/pull/31705#issuecomment-841317965

SiobhyB commented:
Thank you @hypest! I've gone ahead to address each of those points now. For the crash, it was happening because the image was `undefined`, so I've added a check for that around the featured button. Looking forward to hearing any additional feedback!

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -430,6 +434,50 @@ export class ImageEdit extends Component {
+		);
+
+		return (
+			<>
```
Hmm, do we need the (React) fragment here? Looks like we're returning a single component (`PanelBody` in this case) so I wonder if there's a pattern you are following here that opts for using a fragment anyway? 🤔

SiobhyB replied:
Ah, good point, these were only needed when the `androidOnly` conditional was in use here, I overlooked removing them. I've gone ahead to remove them now. 🙇‍♀️

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -489,6 +540,9 @@ export class ImageEdit extends Component {
 					{ this.getLinkSettings( true ) }
 				</PanelBody>
+				{ androidOnly &&
+					typeof attributes.id !== 'undefined' &&
```
Aha, I understand how checking for undefined `id` fixes [the crash](https://github.com/WordPress/gutenberg/pull/31705#issuecomment-841317965), but it's not clear why the code we are in here (the `getInspectorControls` function) should care about this condition which looks too implementation specific. Feels to me that's the responsibility of `getSetFeaturedButton` to care about, especially since the whole PanelBody is managed by `getSetFeaturedButton`.

What do you think about moving the "id undefined" condition inside `getSetFeaturedButton`? While at it, perhaps now the `androidOnly` condition could move there at the function top as well  now that there's going to be one there governing the whole visibility of the set/remove option.


There's one more point to raise here actually: is the undefined state of the `id` a valid state? If yes then indeed it makes sense to cover for it. But, if it's not, then it's a bug we need to fix elsewhere and not try to guard against it here since it will mostly "hide" the issue. What are your thoughts on this?

SiobhyB replied:
The logic outlined in [the previous suggestion](https://github.com/WordPress/gutenberg/pull/31705#discussion_r632567432) to move the `androidOnly` conditional from `getSetFeaturedButton` to here is what led me to also add the check for `undefined` in the same place. Can you help me understand why this check would require a different treatment? The reasoning's not super clear to me just yet. 

> There's one more point to raise here actually: is the undefined state of the id a valid state? If yes then indeed it makes sense to cover for it. But, if it's not, then it's a bug we need to fix elsewhere and not try to guard against it here since it will mostly "hide" the issue. What are your thoughts on this?

`undefined` is a valid state, to my knowledge. An image only has an ID if it's been uploaded to the media library of the site it's being used with. There are valid use cases, such as with page templates, where images are uploaded to another site's media library. Do you have further/different thoughts around that?

hypest replied:
> The logic outlined in the previous suggestion to move the androidOnly conditional from getSetFeaturedButton to here is what led me to also add the check for undefined in the same place. Can you help me understand why this check would require a different treatment? The reasoning's not super clear to me just yet.

Oh sure, sorry. Lemme expand on how I perceive the two conditionals:
1. `androidOnly` is a conditional that can appear anywhere in the code, in an opportunistic way wherever we need to make something platform-specific. In this case, moving it inside `getInspectorControls()` helped un-indent and optimize some code while keeping the logic simple enough: Instead of just "get the set-featured-image button" it became "get the set-featured-image button on Android only".
2. `typeof attributes.id !== 'undefined'` on the other hand is very specific to the featured image logic and captures a very specific detail (don't try to offer the set/remove button if the image doesn't have an id) which the `getInspectorControls()`  shouldn't care about... this conditional is more of a responsibility of a function closer to the feature itself (`getSetFeaturedButton` in this case).

You don't _have_ to move any of conditionals if you don't quite see it fitting. In such a case, I'd probably suggest replacing `typeof attributes.id !== 'undefined'` with a function with a more descriptive name such as "canImageBeFeatured()" which sounds just enough for `getInspectorControls()` to care about, if that makes sense.

No strong feelings either way, let me know if my suggestions are more clear at this stage, thanks!

hypest replied:
> undefined is a valid state, to my knowledge. An image only has an ID if it's been uploaded to the media library of the site it's being used with. There are valid use cases, such as with page templates, where images are uploaded to another site's media library. Do you have further/different thoughts around that?

Perfect, thanks for the context and insight! Makes sense to keep treating a missing `id` as a valid value 👍. Best if we can capture this fact in a comment in the code (perhaps inside a "canImageBeFeatured()" function as I mentioned in the previous comment) as it's good context to have handy. WDYT?

SiobhyB replied:
Great, that's clear, thank you so much for the extended explanation. 🙇‍♀️ I'll work on that!

SiobhyB replied:
I've gone ahead to implement a `canImageBeFeatured` boolean and took the opportunity to use it with the `isFeaturedimage` boolean, as there was some code in it that served the same purpose. Does that look okay to you?

hypest replied:
Looks cool to me, thanks! I've only [added a comment](https://github.com/WordPress/gutenberg/pull/31705#discussion_r634323139) about adding a line comment or similar as there's quite some context packed into that check for `id` and it'd be best if we leave some context near it.

hypest commented:
> A dialog should appear to inform you that the post already has a featured image, with an option to Cancel or confirm your wish to Replace.

Hey @SiobhyB , I tried the "Replace existing" flow and looks like the dialog is gone if a rotate the device while in the process. Is that intended behavior perhaps?

SiobhyB commented:
Thanks for catching that @hypest, definitely not intentional. 🤦‍♀️ I'd copied the code pattern for the dialog from other dialogs in [Android's GutenbergEditorFragment file](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java), such as `showCancelMediaCollectionUploadDialog` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L781) and `showRetryMediaCollectionUploadDialog` [here](https://github.com/wordpress-mobile/WordPress-Android/blob/develop/libs/editor/WordPressEditor/src/main/java/org/wordpress/android/editor/gutenberg/GutenbergEditorFragment.java#L813). After a bit of initial testing, I see that those dialogs are also dismissed/lost upon rotation. I'll do a bit of more research into the background of those dialogs and see if I can work towards a fix.

SiobhyB committed:
Create "canImageBeFeatured" boolean

SiobhyB committed:
Replace some logic in "isFeaturedImage" boolean with new "canImageBeFeatured" boolean

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB committed:
Remove unecessary fragments

SiobhyB commented:
@hypest, I've gone ahead to update the JS side with your suggestions. For the native change, I've created an issue at https://github.com/wordpress-mobile/WordPress-Android/issues/14676 and have referenced a Slack thread where I've asked for feedback on my initial thoughts.

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -496,6 +496,8 @@ export class ImageEdit extends Component {
 			imageDefaultSize,
 		] );
 
+		const canImageBeFeatured = typeof attributes.id !== 'undefined';
```
Let's add a comment to elaborate on what's going on here. I mean, `canImageBeFeatured` has a descriptive name by itself but, why is it that if the image block doesn't have an id it can't become Featured?

SiobhyB replied:
Makes sense! Added a comment now. :)

SiobhyB committed:
Add comment to give context to 'canImageBeFeatured' check

SiobhyB commented:
@hypest, I've gone ahead to implement a new `GutenbergDialogFragment` class [on the Android side](https://github.com/wordpress-mobile/WordPress-Android/pull/14503/files#diff-8889a43bfa80f8522889ae2f69695ac5eda5b29189bb29c9915a9c4825dae6e7) to prevent the dialog being dismissed on screen rotation. The PR isn't 100% ready for review, as I'm still investigating [this bug](https://github.com/wordpress-mobile/WordPress-Android/pull/14503#issuecomment-844950869), but I'd love to get your thoughts on the overall approach of creating that dialog, when you have a chance. 🙇‍♀️

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB commented:
@hypest, just an FYI that the bug I mentioned in the above comment has been fixed now. :)

jd-alexander commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -430,6 +434,48 @@ export class ImageEdit extends Component {
+					setFeaturedButtonStyle,
+					styles.removeFeaturedButton,
+				] }
+				onPress={ () => this.onSetFeatured( 0 ) }
```
@SiobhyB similar to the comment made on the WP Android PR https://github.com/wordpress-mobile/WordPress-Android/pull/14503#discussion_r638422315 what do you think about hardcoding the default media id as a constant for clarity?

SiobhyB replied:
@jd-alexander, that makes sense! I've gone ahead to add the constant in https://github.com/WordPress/gutenberg/pull/31705/commits/5681237eca310bc770ad7bc36ef374d260a8b625, does that seem right to you?

SiobhyB replied:
Just to update: I went ahead to move the constant from the function to the top of the file in https://github.com/WordPress/gutenberg/pull/31705/commits/7f6ad6b758d43e13e2879d9933cc42221a960ddb, as this seemed to match more closely with the pattern I see for other constants. Let me know if there's a better way to declare constants in Gutenberg or if there's something that can be done better here, though!

hypest replied:
Left a [comment on the commit](https://github.com/WordPress/gutenberg/commit/7f6ad6b758d43e13e2879d9933cc42221a960ddb#r51332764) 👍

hypest replied:
Looks like all feedback here has already been addressed and we now have a constant to represent the `0` value. Will mark this convo as resolved :)

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB committed:
Add MEDIA_ID_NO_FEATURED_IMAGE_SET const

With this commit, a MEDIA_ID_NO_FEATURED_IMAGE_SET constant is added to make the reason we pass a zero when a featured image is removed clearer, and also make it easier to change in the future if needed.

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB committed:
Move "MEDIA_ID_NO_FEATURED_IMAGE_SET" constant from within function to top of file

hypest commented on packages/block-library/src/image/edit.native.js:
```diff
@@ -430,6 +436,50 @@ export class ImageEdit extends Component {
+
+		const setFeaturedButton = () => (
+			<BottomSheet.Cell
+				label={ __( 'Remove as Featured Image ' ) }
```
Just noticed, did we get the function names opposite? Looks like the current `setFeaturedButton()` shows a "Remove..." to the user 🤔

SiobhyB replied:
Oh, wow, thanks for spotting that, not sure how that fell through the cracks. 🤦‍♀️ Mixup corrected in https://github.com/WordPress/gutenberg/pull/31705/commits/c309804e10855600f89fa12c473d7b27c2c95afa.

SiobhyB committed:
Correct mixup between "setFeaturedButton" and "removeFeaturedButton" functions

These functions had accidentally been given the oppositive names (that is, the function that removes a featured image was named "setFeaturedButton" and vice versa). This commit corrects that mistake.

SiobhyB committed:
Add 'MEDIA_ID_NO_FEATURED_IMAGE_SET' const to 'image/constants.js'

SiobhyB commented:
👋 @ajitbohra, it seems like [you were added as a reviewer](https://github.com/WordPress/gutenberg/pull/31705#event-4799851881) to this PR following a change to the `image/constants.js` file. Would you be able to let me know whether this was an automatic assignment? I don't recall manually requesting your review, so wanted to double-check.

Most of the changes in the PR itself are on the native side and are already under review by two developers from the mobile team. Would you still need to review the changes in this instance or is it okay to proceed with the current reviewers? I'm not that familiar with the way reviews are assigned in the project, so any insight you can offer around the correct processes would be appreciated. Thanks!

hypest commented:
Tested all the flows [via WPAndroid](https://github.com/wordpress-mobile/WordPress-Android/pull/14503) and works perfect! 🎉

✅ Test Case 1: General Flow
✅ Test Case 2: "Replace Existing" Flow
✅ Test Case 3: "Editor Loads" Flow
✅ Test Case 4: "Post Settings" Flow
✅ Test Case 5: "Replace in Image Block" Flow

hypest commented:
✅ Also verified via WPiOS that the feature is not available there as expected (Android only for the time being, iOS is worked on a different set of PRs).

hypest reviewed (approved):
Looks good to me! Awesome work here @SiobhyB , thanks for persevering and addressing all feedback!

Seems like there are some web E2E test job failing but don't look relevant to the changes here. Perhaps you can update from trunk and see if that fixes them (trunk seems green at the time of writing this).

I'll leave it to the other reviewers of this PR to conclude their rounds and I guess we can start merging the domino! 👏

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

ajitbohra reviewed (approved):
@SiobhyB 

LGTM 👍 

- That's an automatic assignment
- As we already have a code review and 1 approval  its good to merge

jd-alexander reviewed (approved):
Hi @SiobhyB I ran tests after merging in the refactor and all flows work as expected! 🎉

WP Android 

✅ Test Case 1: General Flow
✅ Test Case 2: "Replace Existing" Flow
✅ Test Case 3: "Editor Loads" Flow
✅ Test Case 4: "Post Settings" Flow
✅ Test Case 5: "Replace in Image Block" Flow

WP iOS
✅ Verified that the functionality does not work on iOS and that setting the featured image in Post Settings does not cause the editor to crash.

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting

SiobhyB committed:
Merge branch 'trunk' into update/image-block-with-featured-image-setting
